node: memory manager: fix the metrics tests

ffromani opened the pull request:
#### What type of PR is this?

/kind bug
/kind failing-test

#### What this PR does / why we need it:
Improve the reliability of the tests by saving and recovering the correct kubelet configuration

#### Which issue(s) this PR fixes:
Fixes #123454 

#### Special notes for your reviewer:
This should improve the test signal but not sure it's all we need to fix the tests

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

ffromani commented:
/test pull-crio-cgroupv1-node-e2e-resource-managers
/test pull-kubernetes-e2e-capz-windows-serial-slow
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/sig node

ffromani commented:
/test

ffromani commented:
/test pull-kubernetes-node-arm64-e2e-containerd-serial-ec2
/test pull-kubernetes-node-arm64-ubuntu-serial-gce

ffromani commented:
/test pull-kubernetes-integration

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-kubelet-serial-containerd

ffromani commented:
/test pull-kubernetes-e2e-kind-ipv6

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd

ffromani commented:
latest failure is (surprisingly?) unrelated to the mm metrics test?

ffromani committed:
node: memory manager: fix the mm metrics test

fixes for the memory manager tests by correctly restoring
the kubelet config after each test. We need to do before all
the related tests run, in order to make sure to restore the
correct values.

Add more debug facilities to troubleshoot further failures.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-verify

ffromani commented:
So, we have:

```
> Enter [BeforeEach] [sig-node] Memory Manager Metrics [Serial] [Feature:MemoryManager] - set up framework | framework.go:200 @ 03/04/24 21:04:01.234
STEP: Creating a kubernetes client - k8s.io/kubernetes/test/e2e/framework/framework.go:220 @ 03/04/24 21:04:01.234
STEP: Building a namespace api object, basename memorymanager-metrics - k8s.io/kubernetes/test/e2e/framework/framework.go:259 @ 03/04/24 21:04:01.234
I0304 21:04:01.239148 17026 framework.go:275] Skipping waiting for service account
< Exit [BeforeEach] [sig-node] Memory Manager Metrics [Serial] [Feature:MemoryManager] - set up framework | framework.go:200 @ 03/04/24 21:04:01.239 (5ms)
> Enter [BeforeEach] [sig-node] Memory Manager Metrics [Serial] [Feature:MemoryManager] - k8s.io/kubernetes/test/e2e/framework/metrics/init/init.go:33 @ 03/04/24 21:04:01.239
< Exit [BeforeEach] [sig-node] Memory Manager Metrics [Serial] [Feature:MemoryManager] - k8s.io/kubernetes/test/e2e/framework/metrics/init/init.go:33 @ 03/04/24 21:04:01.239 (0s)
> Enter [BeforeEach] [sig-node] Memory Manager Metrics [Serial] [Feature:MemoryManager] - k8s.io/kubernetes/test/e2e/framework/node/init/init.go:33 @ 03/04/24 21:04:01.239
< Exit [BeforeEach] [sig-node] Memory Manager Metrics [Serial] [Feature:MemoryManager] - k8s.io/kubernetes/test/e2e/framework/node/init/init.go:33 @ 03/04/24 21:04:01.239 (0s)
> Enter [BeforeEach] when querying /metrics - k8s.io/kubernetes/test/e2e_node/memory_manager_metrics_test.go:47 @ 03/04/24 21:04:01.239
STEP: Stopping the kubelet - k8s.io/kubernetes/test/e2e_node/util.go:219 @ 03/04/24 21:04:01.249
I0304 21:04:01.296018 17026 util.go:368] Get running kubelet with systemctl:   UNIT                            LOAD   ACTIVE SUB     DESCRIPTION
  kubelet-20240304T194858.service loaded active running /tmp/node-e2e-20240304T194858/kubelet --kubeconfig /tmp/node-e2e-20240304T194858/kubeconfig --root-dir /var/lib/kubelet --v 4 --config-dir /tmp/node-e2e-20240304T194858/kubelet.conf.d --hostname-override i-0ac0e520963f28681 --container-runtime-endpoint unix:///run/containerd/containerd.sock --config /tmp/node-e2e-20240304T194858/kubelet-config --kernel-memcg-notification=true --feature-gates=DisableKubeletCloudCredentialProviders=true --image-credential-provider-config=/tmp/node-e2e-20240304T194858/credential-provider.yaml --image-credential-provider-bin-dir=/tmp/node-e2e-20240304T194858 --cluster-domain=cluster.local --cgroup-driver=systemd --runtime-cgroups=/system.slice/containerd.service --cgroup-driver=systemd

LOAD   = Reflects whether the unit definition was properly loaded.
ACTIVE = The high-level unit activation state, i.e. generalization of SUB.
SUB    = The low-level unit activation state, values depend on unit type.
1 loaded units listed.
, kubelet-20240304T194858
W0304 21:04:01.402270   17026 util.go:500] Health check on "http://127.0.0.1:10248/healthz" failed, error=Head "http://127.0.0.1:10248/healthz": dial tcp 127.0.0.1:10248: connect: connection refused
STEP: Starting the kubelet - k8s.io/kubernetes/test/e2e_node/util.go:235 @ 03/04/24 21:04:01.406
W0304 21:04:01.509422   17026 util.go:500] Health check on "http://127.0.0.1:10248/healthz" failed, error=Head "http://127.0.0.1:10248/healthz": dial tcp 127.0.0.1:10248: connect: connection refused
< Exit [BeforeEach] when querying /metrics - k8s.io/kubernetes/test/e2e_node/memory_manager_metrics_test.go:47 @ 03/04/24 21:04:06.511 (5.273s)
> Enter [It] should report pinning failures when the memorymanager allocation is known to fail - k8s.io/kubernetes/test/e2e_node/memory_manager_metrics_test.go:102 @ 03/04/24 21:04:06.511
STEP: Creating the test pod which will be rejected for memory request which is too big - k8s.io/kubernetes/test/e2e_node/memory_manager_metrics_test.go:103 @ 03/04/24 21:04:06.511
W0304 21:04:06.514565   17026 warnings.go:70] would violate PodSecurity "restricted:latest": allowPrivilegeEscalation != false (container "memmngrcnt" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container "memmngrcnt" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or container "memmngrcnt" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container "memmngrcnt" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
I0304 21:04:06.516344 17026 memory_manager_metrics_test.go:175] begin listing pods: 2 found
I0304 21:04:06.516371 17026 memory_manager_metrics_test.go:177] kubelet-container-manager-5669/guaranteed994e56e4-0ded-417b-9990-edd5e60b4fd6 node i-0ac0e520963f28681 status Succeeded QoS Guaranteed message  reason  (1 container statuses recorded)
I0304 21:04:06.516379 17026 memory_manager_metrics_test.go:179] 	Container guaranteed994e56e4-0ded-417b-9990-edd5e60b4fd6 ready: false, restart count 0
I0304 21:04:06.516392 17026 memory_manager_metrics_test.go:177] memorymanager-metrics-7056/memmngrpodqxfln node i-0ac0e520963f28681 status Pending QoS Guaranteed message  reason  (0 container statuses recorded)
I0304 21:04:06.516401 17026 memory_manager_metrics_test.go:183] end listing pods: 2 found
STEP: Checking the memorymanager metrics right after the kubelet restart, with pod failed to admit - k8s.io/kubernetes/test/e2e_node/memory_manager_metrics_test.go:117 @ 03/04/24 21:04:06.516
STEP: Giving the Kubelet time to start up and produce metrics - k8s.io/kubernetes/test/e2e_node/memory_manager_metrics_test.go:128 @ 03/04/24 21:04:06.516
STEP: getting Kubelet metrics from the metrics API - k8s.io/kubernetes/test/e2e_node/cpu_manager_metrics_test.go:167 @ 03/04/24 21:04:06.516
STEP: getting Kubelet metrics from the metrics API - k8s.io/kubernetes/test/e2e_node/cpu_manager_metrics_test.go:167 @ 03/04/24 21:04:21.536
STEP: getting Kubelet metrics from the metrics API - k8s.io/kubernetes/test/e2e_node/cpu_manager_metrics_test.go:167 @ 03/04/24 21:04:36.545
STEP: getting Kubelet metrics from the metrics API - k8s.io/kubernetes/test/e2e_node/cpu_manager_metrics_test.go:167 @ 03/04/24 21:04:51.556
[FAILED] Timed out after 60.000s.
Expected
    <string>: KubeletMetrics
to match keys: {
."kubelet_memory_manager_pinning_requests_total"[]:
	Expected
	    <string>: Sample
	to match fields: {
	.Value:
		Expected
		    <model.SampleValue>: 2
		to be ==
		    <int>: 1
	}
	
}
In [It] at: k8s.io/kubernetes/test/e2e_node/memory_manager_metrics_test.go:129 @ 03/04/24 21:05:06.516
< Exit [It] should report pinning failures when the memorymanager allocation is known to fail - k8s.io/kubernetes/test/e2e_node/memory_manager_metrics_test.go:102 @ 03/04/24 21:05:06.516 (1m0.005s)
> Enter [DeferCleanup (Each)] when querying /metrics - k8s.io/kubernetes/test/e2e_node/memory_manager_metrics_test.go:72 @ 03/04/24 21:05:06.516
STEP: Stopping the kubelet - k8s.io/kubernetes/test/e2e_node/util.go:219 @ 03/04/24 21:05:06.523
I0304 21:05:06.548492 17026 util.go:368] Get running kubelet with systemctl:   UNIT                            LOAD   ACTIVE SUB     DESCRIPTION
  kubelet-20240304T194858.service loaded active running /tmp/node-e2e-20240304T194858/kubelet --kubeconfig /tmp/node-e2e-20240304T194858/kubeconfig --root-dir /var/lib/kubelet --v 4 --config-dir /tmp/node-e2e-20240304T194858/kubelet.conf.d --hostname-override i-0ac0e520963f28681 --container-runtime-endpoint unix:///run/containerd/containerd.sock --config /tmp/node-e2e-20240304T194858/kubelet-config --kernel-memcg-notification=true --feature-gates=DisableKubeletCloudCredentialProviders=true --image-credential-provider-config=/tmp/node-e2e-20240304T194858/credential-provider.yaml --image-credential-provider-bin-dir=/tmp/node-e2e-20240304T194858 --cluster-domain=cluster.local --cgroup-driver=systemd --runtime-cgroups=/system.slice/containerd.service --cgroup-driver=systemd

LOAD   = Reflects whether the unit definition was properly loaded.
ACTIVE = The high-level unit activation state, i.e. generalization of SUB.
SUB    = The low-level unit activation state, values depend on unit type.
1 loaded units listed.
, kubelet-20240304T194858
W0304 21:05:06.653396   17026 util.go:500] Health check on "http://127.0.0.1:10248/healthz" failed, error=Head "http://127.0.0.1:10248/healthz": read tcp 127.0.0.1:44000->127.0.0.1:10248: read: connection reset by peer
STEP: Starting the kubelet - k8s.io/kubernetes/test/e2e_node/util.go:235 @ 03/04/24 21:05:06.657
W0304 21:05:06.739242   17026 util.go:500] Health check on "http://127.0.0.1:10248/healthz" failed, error=Head "http://127.0.0.1:10248/healthz": dial tcp 127.0.0.1:10248: connect: connection refused
< Exit [DeferCleanup (Each)] when querying /metrics - k8s.io/kubernetes/test/e2e_node/memory_manager_metrics_test.go:72 @ 03/04/24 21:05:11.745 (5.229s)
> Enter [DeferCleanup (Each)] [sig-node] Memory Manager Metrics [Serial] [Feature:MemoryManager] - k8s.io/kubernetes/test/e2e/framework/node/init/init.go:34 @ 03/04/24 21:05:11.745
I0304 21:05:11.745748 17026 helper.go:121] Waiting up to 7m0s for all (but 0) nodes to be ready
< Exit [DeferCleanup (Each)] [sig-node] Memory Manager Metrics [Serial] [Feature:MemoryManager] - k8s.io/kubernetes/test/e2e/framework/node/init/init.go:34 @ 03/04/24 21:05:11.747 (1ms)
> Enter [DeferCleanup (Each)] [sig-node] Memory Manager Metrics [Serial] [Feature:MemoryManager] - k8s.io/kubernetes/test/e2e/framework/metrics/init/init.go:35 @ 03/04/24 21:05:11.747
< Exit [DeferCleanup (Each)] [sig-node] Memory Manager Metrics [Serial] [Feature:MemoryManager] - k8s.io/kubernetes/test/e2e/framework/metrics/init/init.go:35 @ 03/04/24 21:05:11.747 (0s)
> Enter [DeferCleanup (Each)] [sig-node] Memory Manager Metrics [Serial] [Feature:MemoryManager] - dump namespaces | framework.go:218 @ 03/04/24 21:05:11.747
STEP: dump namespace information after failure - k8s.io/kubernetes/test/e2e/framework/framework.go:297 @ 03/04/24 21:05:11.747
STEP: Collecting events from namespace "memorymanager-metrics-7056". - k8s.io/kubernetes/test/e2e/framework/debug/dump.go:42 @ 03/04/24 21:05:11.747
STEP: Found 1 events. - k8s.io/kubernetes/test/e2e/framework/debug/dump.go:46 @ 03/04/24 21:05:11.748
I0304 21:05:11.748360 17026 dump.go:53] At 2024-03-04 21:04:06 +0000 UTC - event for memmngrpodqxfln: {kubelet i-0ac0e520963f28681} UnexpectedAdmissionError: Allocate failed due to [memorymanager] failed to get the default NUMA affinity, no NUMA nodes with enough memory is available, which is unexpected
I0304 21:05:11.749206 17026 resource.go:168] POD  NODE  PHASE  GRACE  CONDITIONS
I0304 21:05:11.749229 17026 resource.go:178] 
I0304 21:05:11.750364 17026 dump.go:109] 
Logging node info for node i-0ac0e520963f28681
I0304 21:05:11.751605 17026 dump.go:114] Node Info: &Node{ObjectMeta:{i-0ac0e520963f28681    0179d058-d76f-4236-a5fe-fea5c41712ee 4114 0 2024-03-04 19:49:10 +0000 UTC <nil> <nil> map[beta.kubernetes.io/arch:amd64 beta.kubernetes.io/os:linux kubernetes.io/arch:amd64 kubernetes.io/hostname:i-0ac0e520963f28681 kubernetes.io/os:linux] map[volumes.kubernetes.io/controller-managed-attach-detach:true] [] [] [{kubelet Update v1 2024-03-04 19:49:10 +0000 UTC FieldsV1 {"f:metadata":{"f:annotations":{".":{},"f:volumes.kubernetes.io/controller-managed-attach-detach":{}},"f:labels":{".":{},"f:beta.kubernetes.io/arch":{},"f:beta.kubernetes.io/os":{},"f:kubernetes.io/hostname":{}}}} } {kubelet Update v1 2024-03-04 21:05:06 +0000 UTC FieldsV1 {"f:metadata":{"f:labels":{"f:kubernetes.io/arch":{},"f:kubernetes.io/os":{}}},"f:status":{"f:allocatable":{"f:cpu":{},"f:ephemeral-storage":{},"f:example.com/resource":{},"f:hugepages-2Mi":{},"f:memory":{}},"f:capacity":{"f:example.com/resource":{},"f:hugepages-2Mi":{}},"f:conditions":{"k:{\"type\":\"DiskPressure\"}":{"f:lastHeartbeatTime":{},"f:lastTransitionTime":{},"f:message":{},"f:reason":{},"f:status":{}},"k:{\"type\":\"MemoryPressure\"}":{"f:lastHeartbeatTime":{}},"k:{\"type\":\"PIDPressure\"}":{"f:lastHeartbeatTime":{}},"k:{\"type\":\"Ready\"}":{"f:lastHeartbeatTime":{},"f:lastTransitionTime":{},"f:message":{},"f:reason":{},"f:status":{}}},"f:images":{},"f:nodeInfo":{"f:containerRuntimeVersion":{}}}} status}]},Spec:NodeSpec{PodCIDR:,DoNotUseExternalID:,ProviderID:,Unschedulable:false,Taints:[]Taint{},ConfigSource:nil,PodCIDRs:[],},Status:NodeStatus{Capacity:ResourceList{cpu: {{2 0} {<nil>} 2 DecimalSI},ephemeral-storage: {{53607378944 0} {<nil>} 52350956Ki BinarySI},example.com/resource: {{0 0} {<nil>} 0 DecimalSI},hugepages-1Gi: {{0 0} {<nil>} 0 DecimalSI},hugepages-2Mi: {{0 0} {<nil>} 0 DecimalSI},memory: {{8103153664 0} {<nil>} 7913236Ki BinarySI},pods: {{110 0} {<nil>} 110 DecimalSI},},Allocatable:ResourceList{cpu: {{2 0} {<nil>} 2 DecimalSI},ephemeral-storage: {{48246640970 0} {<nil>} 48246640970 DecimalSI},example.com/resource: {{0 0} {<nil>} 0 DecimalSI},hugepages-1Gi: {{0 0} {<nil>} 0 DecimalSI},hugepages-2Mi: {{0 0} {<nil>} 0 DecimalSI},memory: {{7841009664 0} {<nil>} 7657236Ki BinarySI},pods: {{110 0} {<nil>} 110 DecimalSI},},Phase:,Conditions:[]NodeCondition{NodeCondition{Type:MemoryPressure,Status:False,LastHeartbeatTime:2024-03-04 21:05:06 +0000 UTC,LastTransitionTime:2024-03-04 19:49:10 +0000 UTC,Reason:KubeletHasSufficientMemory,Message:kubelet has sufficient memory available,},NodeCondition{Type:DiskPressure,Status:False,LastHeartbeatTime:2024-03-04 21:05:06 +0000 UTC,LastTransitionTime:2024-03-04 20:01:53 +0000 UTC,Reason:KubeletHasNoDiskPressure,Message:kubelet has no disk pressure,},NodeCondition{Type:PIDPressure,Status:False,LastHeartbeatTime:2024-03-04 21:05:06 +0000 UTC,LastTransitionTime:2024-03-04 19:49:10 +0000 UTC,Reason:KubeletHasSufficientPID,Message:kubelet has sufficient PID available,},NodeCondition{Type:Ready,Status:True,LastHeartbeatTime:2024-03-04 21:05:06 +0000 UTC,LastTransitionTime:2024-03-04 20:41:59 +0000 UTC,Reason:KubeletReady,Message:kubelet is posting ready status,},},Addresses:[]NodeAddress{NodeAddress{Type:InternalIP,Address:172.31.91.175,},NodeAddress{Type:Hostname,Address:i-0ac0e520963f28681,},},DaemonEndpoints:NodeDaemonEndpoints{KubeletEndpoint:DaemonEndpoint{Port:10250,},},NodeInfo:NodeSystemInfo{MachineID:ec2ac97c04ac656aae9f9169a562abc2,SystemUUID:ec2ac97c-04ac-656a-ae9f-9169a562abc2,BootID:83a00488-fb56-4d28-a2d4-5aab2106d304,KernelVersion:6.1.77-99.164.amzn2023.x86_64,OSImage:Amazon Linux 2023,ContainerRuntimeVersion:containerd://2.0.0-beta.2-222-g9a2b85561,KubeletVersion:v1.30.0-alpha.3.375+4ddf96c672283f,KubeProxyVersion:v1.30.0-alpha.3.375+4ddf96c672283f,OperatingSystem:linux,Architecture:amd64,},Images:[]ContainerImage{ContainerImage{Names:[registry.k8s.io/e2e-test-images/perl@sha256:c613344cdd31c5055961b078f831ef9d9199fc9111efe6e81bea3f00d78bd979 registry.k8s.io/e2e-test-images/perl:5.26],SizeBytes:325397679,},ContainerImage{Names:[registry.k8s.io/e2e-test-images/node-perf/tf-wide-deep@sha256:91ab3b5ee22441c99370944e2e2cb32670db62db433611b4e3780bdee6a8e5a1 registry.k8s.io/e2e-test-images/node-perf/tf-wide-deep:1.3],SizeBytes:213443508,},ContainerImage{Names:[registry.k8s.io/e2e-test-images/volume/nfs@sha256:c3e380bb5a92095288b4dd586a250c560612afbabf53fb2a057e99fff75f038f registry.k8s.io/e2e-test-images/volume/nfs:1.4],SizeBytes:95831867,},ContainerImage{Names:[registry.k8s.io/node-problem-detector/node-problem-detector@sha256:d65a5c35dc7e948ff6d0bbe15b5e272c55790acbc9db4963ad91dc9939b4e293 registry.k8s.io/node-problem-detector/node-problem-detector:v0.8.13],SizeBytes:57958233,},ContainerImage{Names:[registry.k8s.io/etcd@sha256:44a8e24dcbba3470ee1fee21d5e88d128c936e9b55d4bc51fbef8086f8ed123b registry.k8s.io/etcd:3.5.12-0],SizeBytes:57236178,},ContainerImage{Names:[registry.k8s.io/e2e-test-images/agnhost@sha256:cc249acbd34692826b2b335335615e060fdb3c0bca4954507aa3a1d1194de253 registry.k8s.io/e2e-test-images/agnhost:2.47],SizeBytes:52529911,},ContainerImage{Names:[registry.k8s.io/e2e-test-images/httpd@sha256:148b022f5c5da426fc2f3c14b5c0867e58ef05961510c84749ac1fddcb0fef22 registry.k8s.io/e2e-test-images/httpd:2.4.38-4],SizeBytes:40764257,},ContainerImage{Names:[registry.k8s.io/e2e-test-images/node-perf/npb-is@sha256:8285539c79625b192a5e33fc3d21edc1a7776fb9afe15fae3b5037a7a8020839 registry.k8s.io/e2e-test-images/node-perf/npb-is:1.2],SizeBytes:39706325,},ContainerImage{Names:[registry.k8s.io/e2e-test-images/node-perf/npb-ep@sha256:90b5cfc5451428aad4dd6af9960640f2506804d35aa05e83c11bf0a46ac318c8 registry.k8s.io/e2e-test-images/node-perf/npb-ep:1.2],SizeBytes:39705373,},ContainerImage{Names:[gcr.io/cadvisor/cadvisor@sha256:e6c562b5e983f13624898b5b6a902c71999580dc362022fc327c309234c485d7 gcr.io/cadvisor/cadvisor:v0.47.2],SizeBytes:33756591,},ContainerImage{Names:[registry.k8s.io/e2e-test-images/sample-device-plugin@sha256:333840c95378b183bb004a858fd456a2db0116e285b45f57c42e25a17fe130b5 registry.k8s.io/e2e-test-images/sample-device-plugin:1.7],SizeBytes:17861073,},ContainerImage{Names:[registry.k8s.io/e2e-test-images/nonroot@sha256:ee9f50b3c64b174d296d91ca9f69a914ac30e59095dfb462b2b518ad28a63655 registry.k8s.io/e2e-test-images/nonroot:1.4],SizeBytes:17747885,},ContainerImage{Names:[docker.io/nfvpe/sriov-device-plugin@sha256:518499ed631ff84b43153b8f7624c1aaacb75a721038857509fe690abdf62ddb docker.io/nfvpe/sriov-device-plugin:v3.1],SizeBytes:12638340,},ContainerImage{Names:[registry.k8s.io/e2e-test-images/nginx@sha256:5c99cf6a02adda929b10321dbf4ecfa00d87be9ba4fb456006237d530ab4baa1 registry.k8s.io/e2e-test-images/nginx:1.14-4],SizeBytes:6978614,},ContainerImage{Names:[registry.k8s.io/nvidia-gpu-device-plugin@sha256:4b036e8844920336fa48f36edeb7d4398f426d6a934ba022848deed2edbf09aa],SizeBytes:6819465,},ContainerImage{Names:[registry.k8s.io/e2e-test-images/ipc-utils@sha256:647d092bada3b46c449d875adf31d71c1dd29c244e9cca6a04fddf9d6bcac136 registry.k8s.io/e2e-test-images/ipc-utils:1.3],SizeBytes:4013020,},ContainerImage{Names:[registry.k8s.io/e2e-test-images/nonewprivs@sha256:8ac1264691820febacf3aea5d152cbde6d10685731ec14966a9401c6f47a68ac registry.k8s.io/e2e-test-images/nonewprivs:1.3],SizeBytes:3263463,},ContainerImage{Names:[registry.k8s.io/e2e-test-images/busybox@sha256:a9155b13325b2abef48e71de77bb8ac015412a566829f621d06bfae5c699b1b9 registry.k8s.io/e2e-test-images/busybox:1.36.1-1],SizeBytes:2223659,},ContainerImage{Names:[registry.k8s.io/pause@sha256:7031c1b283388d2c2e09b57badb803c05ebed362dc88d84b480cc47f72a21097 registry.k8s.io/pause:3.9],SizeBytes:321520,},},VolumesInUse:[],VolumesAttached:[]AttachedVolume{},Config:nil,},}
I0304 21:05:11.751631 17026 dump.go:116] 
Logging kubelet events for node i-0ac0e520963f28681
I0304 21:05:11.752595 17026 dump.go:121] 
Logging pods the kubelet thinks is on node i-0ac0e520963f28681
W0304 21:05:11.758398   17026 metrics_grabber.go:111] Can't find any pods in namespace kube-system to grab metrics from
I0304 21:05:11.769542 17026 kubelet_metrics.go:206] 
Latency metrics for node i-0ac0e520963f28681
END STEP: dump namespace information after failure - k8s.io/kubernetes/test/e2e/framework/framework.go:297 @ 03/04/24 21:05:11.769 (22ms)
< Exit [DeferCleanup (Each)] [sig-node] Memory Manager Metrics [Serial] [Feature:MemoryManager] - dump namespaces | framework.go:218 @ 03/04/24 21:05:11.769 (22ms)
> Enter [DeferCleanup (Each)] [sig-node] Memory Manager Metrics [Serial] [Feature:MemoryManager] - tear down framework | framework.go:215 @ 03/04/24 21:05:11.769
STEP: Destroying namespace "memorymanager-metrics-7056" for this suite. - k8s.io/kubernetes/test/e2e/framework/framework.go:360 @ 03/04/24 21:05:11.769
< Exit [DeferCleanup (Each)] [sig-node] Memory Manager Metrics [Serial] [Feature:MemoryManager] - tear down framework | framework.go:215 @ 03/04/24 21:05:11.771 (2ms)
```

I think it's better and safer to relax the test conditions instead of looking for exact matches. Fixing in a little while.

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-kubelet-serial-containerd

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-kubelet-serial-containerd

ffromani commented:
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-node-kubelet-serial-containerd

swatisehgal commented:
/triage accepted

ffromani commented on test/e2e_node/memory_manager_metrics_test.go:
```diff
@@ -101,23 +108,28 @@ var _ = SIGDescribe("Memory Manager Metrics", framework.WithSerial(), feature.Me
+				}),
+			)
+
+			printAllPodsOnNode(ctx, f.ClientSet, framework.TestContext.NodeName)
 
 			// we updated the kubelet config in BeforeEach, so we can assume we start fresh.
 			// being [Serial], we can also assume noone else but us is running pods.
 			ginkgo.By("Checking the memorymanager metrics right after the kubelet restart, with pod failed to admit")
 
 			matchResourceMetrics := gstruct.MatchKeys(gstruct.IgnoreExtras, gstruct.Keys{
 				"kubelet_memory_manager_pinning_requests_total": gstruct.MatchAllElements(nodeID, gstruct.Elements{
-					"": timelessSample(1),
+					"": memoryTimelessSample(1),
 				}),
 				"kubelet_memory_manager_pinning_errors_total": gstruct.MatchAllElements(nodeID, gstruct.Elements{
-					"": timelessSample(1),
+					"": memoryTimelessSample(1),
```
We do indeed have leaked pods in some (many?) runs: https://github.com/kubernetes/kubernetes/pull/123468#issuecomment-1977977609

I can't pinpoint yet why we have unexpected pods, but being the case the tests run periodically/on-demand (vs against each PR), we will always lagging behind and we will always need to chase other e2e tests to make sure they do the cleanup properly. Since we are already severely undercapacity I don't think this would help.
I do agree the current check is too relaxed (but OTOH I also believe the former was too strict).
A better check would be compare total and errors, and ensure that in our testbed total > errors, I'll be working on that.

Tal-or replied:
> we will always lagging behind and we will always need to chase other e2e tests to make sure they do the cleanup properly

Since the tests run in serial fashion I don't think it would be too complex to delete all running guaranteed QoS pods before we start.
We should list all pods, filter by QoS and delete shouldn't be too expensive IMO

ffromani replied:
true, but still this is a workaround, which makes the ecosystem slightly worse, despite good intention. The proper fix would be to fix the leaking tests instead of piling workarounds.

Tal-or replied:
I see, does the name of the rebellious pod shade some light about his origin?

ffromani replied:
Yes, it's `test/e2e_node/container_manager_test.go`. I'm against adding checks in _this_ test because failing the preconditions will make _this_ test red, kinda hiding the misbehaving part.

Tal-or replied:
I agree. And I presume cleaning the `test/e2e_node/container_manager_test.go` properly is an option but it might bite us again with another careless cleanup in the future

swatisehgal commented:
/priority important-longterm

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-kubelet-serial-containerd

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-kubelet-serial-containerd

ffromani committed:
e2e: node: ensure no pod leaks in the container_manager test

During the debugging of https://github.com/kubernetes/kubernetes/pull/123468
it became quite evident there are unexpected pods, leftovers from
the container_manager_test. But we need stronger isolation among test
to have good signal, so we add these safeguards (xref:
https://github.com/kubernetes/kubernetes/pull/123468#issuecomment-1977977609
)

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-kubelet-serial-containerd

ffromani commented:
the MM metrics tests are consistently passing (or at least much less flaky). We're getting unrelated failures, which should be tracked separately

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-kubelet-serial-containerd

pacoxu commented:
https://testgrid.k8s.io/sig-node-release-blocking#node-kubelet-serial-containerd
This fix a flake in sig-node-release-blocking.

/cc @bart0sh @kannon92 
/assign @SergeyKanzhelev @klueska @mrunalp

kannon92 commented:
@swatisehgal could you take a look?

swatisehgal reviewed (commented):
Thanks Francesco for all the work on this.

I see CI is green and happy and I am trying really hard to be happy with the changes here but there are still some loose ends that I can't fully make sense of so please bear with me :) 

Our first test is to check the state of memory manager metric right after kubelet restart where we have a **strict** check and basically want to ensure that both "kubelet_memory_manager_pinning_requests_total" and "kubelet_memory_manager_pinning_errors_total" have zero values which makes perfect sense to me.

If the environment is not clean, the test fails because of this strict first check and rightly so! However, in the subsequent tests (one where pod fails at admission and another one where it succeeds) we lax the checks.

Now what is not clear to me is that if we ensure that our starting point is a clean environment and we have no leaked pod why do we need to make the other checks less strict and how does that help?

Let's take the second test (`should report pinning failures when the memorymanager allocation is known to fail`) as an example here, we have created **1 pod** after kubelet restarts and yet are okay if the metric indicates there are more than 1 pod requests and failures [value > = 1](https://github.com/kubernetes/kubernetes/pull/123468/commits/6691032fec98d181a74df77513a1adec2f70e33c#diff-6f3fbea74d4f9dfbdb8ee8b6f9415a1a266750ff6ebc7946ae496f18e1687fbcR139). Both the metrics incrementing by 1 and only 1 is indicative of the fact that the metrics are being correctly updated and if that is not the case, it means that they are not getting updated as expected and the test **should** fail. I am worried that in the future these relaxed checks could result in false positives and impression of a happy and healthy CI when that is not the case.  

Also I don't fully understand the comment about not being able to afford strict checks due to capacity constraints. Let's assume that some tests are added in the future and fail to clean the environment properly. In that case, the CI would end up turning RED because of the failure of the first strict check and someone from the community would _have_ to investigate the issue. How do the laxed checks help us here?

ffromani commented:
Thanks @swatisehgal for your deep and thoughtful review, this is very very appreciated. Unfortunately I need to handle other activities now but I will address your points ASAP. Just wanted to give a quick and deserved acknowledgement.

ffromani commented on test/e2e_node/memory_manager_metrics_test.go:
```diff
@@ -80,19 +85,36 @@ var _ = SIGDescribe("Memory Manager Metrics", framework.WithSerial(), feature.Me
 			gomega.Eventually(getKubeletMetrics, 1*time.Minute, 15*time.Second).WithContext(ctx).Should(matchResourceMetrics)
 			ginkgo.By("Ensuring the metrics match the expectations a few more times")
 			gomega.Consistently(getKubeletMetrics, 1*time.Minute, 15*time.Second).WithContext(ctx).Should(matchResourceMetrics)
+
+			values, err := getKubeletMetrics(ctx)
+			framework.ExpectNoError(err, "error getting the kubelet metrics for sanity check")
+			err = validateMetrics(
+				values,
+				"kubelet_memory_manager_pinning_requests_total",
+				"kubelet_memory_manager_pinning_errors_total",
+				func(totVal, errVal float64) error {
+					if int64(totVal) != int64(errVal) {
+						return fmt.Errorf("expected total requests equal to total errors")
+					}
+					return nil
+				},
+			)
+			framework.ExpectNoError(err, "error validating the kubelet metrics between each other")
```
true, overzealous copy/paste. Fixing.

Tal-or commented:
I'll add my 2 cent here and say that maybe we should consider a proper cleanup before the test run than making the tests less stricter.

Having a cleanup on our side hiding the problem - true but same claim goes for laxing the tests.
Also while fixing the actual culprit is the right approach it doesn't guaranteed we won't hit the same issue in the future by a different test.

Having a proper cleanup from our side is indeed a workaround but it at least guarantees that we won't face the same issue again.

ffromani commented:
> Having a cleanup on our side hiding the problem - true but same claim goes for laxing the tests. Also while fixing the actual culprit is the right approach it doesn't guaranteed we won't hit the same issue in the future by a different test.

I get what you mean, and this is probably something we should try. However there's a fine line on which we can relax the test without making it incorrect, which was the spot I was aiming for.
 
> Having a proper cleanup from our side is indeed a workaround but it at least guarantees that we won't face the same issue again.

That's a good point, so I'll probably try this approach

ffromani commented:
added a precondition check in `BeforeEach` and restore the strict check

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-kubelet-serial-containerd

ffromani commented:
/test pull-kubernetes-e2e-capz-windows-serial-slow
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-kubelet-serial-containerd

Tal-or commented:
IIUC we made sure no leak in there are no leaks in the container_manager test, but since the tests are not running in a serial fashion, what guarantees we won't hit the same issue again, by a different test this time?

ffromani commented:
> IIUC we made sure no leak in there are no leaks in the container_manager test, but since the tests are not running in a serial fashion, what guarantees we won't hit the same issue again, by a different test this time?

Nothing will, only reviews, careful checks and fixing once we notify. This was what I tried to convey in https://github.com/kubernetes/kubernetes/pull/123468#discussion_r1512741529 and https://github.com/kubernetes/kubernetes/pull/123468#issuecomment-2007399155

ffromani commented:
swap-related failures are a known issue addressed in https://github.com/kubernetes/kubernetes/pull/125402

ffromani commented:
> > > IIUC we made sure no leak in there are no leaks in the container_manager test, but since the tests are not running in a serial fashion, what guarantees we won't hit the same issue again, by a different test this time?
> > 
> > 
> > Nothing will, only reviews, careful checks and fixing once we notice, so after the breakage. This was what I tried to convey in [#123468 (comment)](https://github.com/kubernetes/kubernetes/pull/123468#discussion_r1512741529) and [#123468 (comment)](https://github.com/kubernetes/kubernetes/pull/123468#issuecomment-2007399155)
> 
> The current fix is great and actually solving the problem (without cutting corners), but I would like to gently push back about why not deleting any workloads before the metric test starts? This will guaranteed we won't hit this issue again.

So, we want to add a cleanup code in `BeforeEach`?  This is likely to ensure _these_ tests won't fail (because will run against clean environment), but will actually hide other buggy and leaking tests.

As I see it there are 3 ways:
- ensure no new or existing test leak :)
- check the preconditions and fail if not met. This makes the issues evident but moves the blame on the messenger (this is what we have now)
- enforce the preconditions. This makes the current tests harder to fail, but will hide buggy tests

It's all about choosing between evils it seems

ffromani commented:
> > > > IIUC we made sure no leak in there are no leaks in the container_manager test, but since the tests are not running in a serial fashion, what guarantees we won't hit the same issue again, by a different test this time?
> > > 
> > > 
> > > Nothing will, only reviews, careful checks and fixing once we notice, so after the breakage. This was what I tried to convey in [#123468 (comment)](https://github.com/kubernetes/kubernetes/pull/123468#discussion_r1512741529) and [#123468 (comment)](https://github.com/kubernetes/kubernetes/pull/123468#issuecomment-2007399155)
> > 
> > 
> > The current fix is great and actually solving the problem (without cutting corners), but I would like to gently push back about why not deleting any workloads before the metric test starts? This will guaranteed we won't hit this issue again.
> 
> So, we want to add a cleanup code in `BeforeEach`? This is likely to ensure _these_ tests won't fail (because will run against clean environment), but will actually hide other buggy and leaking tests.
> 
> As I see it there are 3 ways:
> 
>     * ensure no new or existing test leak :)
> 
>     * check the preconditions and fail if not met. This makes the issues evident but moves the blame on the messenger (this is what we have now)
> 
>     * enforce the preconditions. This makes the current tests harder to fail, but will hide buggy tests
> 
> 
> It's all about choosing between evils it seems

I think I may have a solution, I'll post a separate PR shortly

Tal-or commented:
Thank you for working on this, considering the fact that now we have strong evident in case of pod leakage (https://github.com/kubernetes/kubernetes/pull/125434) LGTM

SergeyKanzhelev commented on test/e2e_node/container_manager_test.go:
```diff
@@ -100,7 +100,32 @@ var _ = SIGDescribe("Container Manager Misc", framework.WithSerial(), func() {
+						}
+					}
+
+					deletePodSyncByName(ctx, f, testPod.Name)
```
what if test pod was not assgned yet? Can we add a check for this please

ffromani replied:
Right, thanks. Fixed.

SergeyKanzhelev commented on test/e2e_node/container_manager_test.go:
```diff
@@ -100,7 +100,32 @@ var _ = SIGDescribe("Container Manager Misc", framework.WithSerial(), func() {
+							},
+						})
+						framework.ExpectNoError(err)
+						framework.Logf("Running containers:")
```
I am trying to remember, isn't it being dumped for failed tests already?

ffromani replied:
not that I recall, this is why I added this code, to find out the leaked pod

ffromani commented on test/e2e_node/container_manager_test.go:
```diff
@@ -100,7 +100,32 @@ var _ = SIGDescribe("Container Manager Misc", framework.WithSerial(), func() {
+								State: runtimeapi.ContainerState_CONTAINER_RUNNING,
+							},
+						})
+						framework.ExpectNoError(err)
```
right, fixing

ffromani commented:
/retest-required

SergeyKanzhelev commented on test/e2e_node/container_manager_test.go:
```diff
@@ -78,6 +78,26 @@ func validateOOMScoreAdjSettingIsInRange(pid int, expectedMinOOMScoreAdj, expect
 	return nil
 }
 
+func dumpRunningContainer(ctx context.Context) error {
```
as discussed it may be nice to generalize this in future. lgtm for now.

SergeyKanzhelev reviewed (approved):
/lgtm
/approve
